kong/tools/wrpc.lua

```diff
@@ -645,41 +631,22 @@ function wrpc_peer:get_response(req_id)
 end
 
 
-local function send_payload_to_channel(self, payload)
-  assert(self.channel:post(self.encode("wrpc.PayloadV1", payload)))
-end
-
-local function remote_call(self, name, ...)
-  self:call(name, ...)
-
-  local msg = assert(self.channel:get())
-  local payload_back = assert(self.decode("wrpc.PayloadV1", msg))
-
-  if payload_back.mtype == "MESSAGE_TYPE_ERROR" then
-    return nil, payload_back.error.description
+local wrpc_remote_peer = {}
+wrpc_remote_peer.__index = function(self, name)
+  local c = channels[self.channel_name]
+  if not c then
+    ngx.log(ngx.ERR, "channel \"", self.channel_name, "\" does not exist when trying to call ", name, " from client")
   end
-
-  if payload_back.mtype == "MESSAGE_TYPE_RPC" then
-    local rpc = self.service:get_method(payload_back.svc_id, payload_back.rpc_id)
-    return decodearray(self.decode, rpc.output_type, payload_back.payloads)
+  if c[name] then
+    return c[name]
   end
-
-  return nil, "unknown message type"
 end
 
-local function remote_close(self)
-  self.closing = true
+function wrpc.new_remote_client(channel_name)
+  local ret = setmetatable({
+    channel_name = channel_name
+  }, wrpc_remote_peer)
+  return ret
 end
```

Comment:
as i understand it, the `channels` table is just a registry of peers.  given that this won't work across workers, what's the need for the wrapper object? why not just pick the "real" peer from the registry?

Reply:
For the first question: we choose to not support access from other workers (currently all DP/CP communication is done within a privileged worker).
For the second question, if the connection is lost, the config service is aware of the reconnection (and handles it), but the user of `remote_peer` won't know that. It's important for `remote_peer` to work with reconnection.

Reply:
I think eventually we will make the config service in the same position as other services, and the reconnection should be handled by a separate thread, or automatically by the wRPC implementation. But for now I'll just leave it as it is.

Reply:
my question is: why use a wrapper object at all?  since it's limited to the same worker, it should just use the same peer object

Reply:
> my question is: why use a wrapper object at all? since it's limited to the same worker, it should just use the same peer object

The wrapper checks the connections' existence, so users of the "remote" connection don't need to check for existence every time. This may be not that important but it's useful. (Maybe the peer will be able to handle reconnection and we can always use the "real" peer, and then we don't need this anymore, but we will have a similar interface.)

Reply:
Or rather we can always return the same object for the same name of channel, like this:
```lua
local channels = {}
function wrpc.new_peer(name, conn, service, opts)
  local ret = channels[name] or {}
  ret.conn = conn or ret.conn,
  ret.service = service or ret.service
  if opts  then
    merge(ret, opts)
  end
  channels[name] = ret
  return setmetatable(ret, wrpc_peer)
end
```

Reply:
And I think maybe we should decouple the "channel" concept.
For example, use `kong.dp_connection` and `kong.cp_connection` to get the wRPC connection to the remote peer.